### PR TITLE
Python: more additional taint steps

### PR DESCRIPTION
**Pull request by RasmusWL**
The taint steps defined here are fairly rough, and are only using syntactical matching. The idea is that this should be good enough as a first step, and then we can focus on more interesting things such as using TypeTracking :wink: 

This implementation is very generous in handing out taint. As in #4124 we don't make any attempt to limit taint propagation based on types, so since we want `<tainted dict>.get(<key>)` to propagate taint, we also propagate taint for `<any tainted obj>.get()`. I'm not too sure whether this turns out to be a problem or not.

The one thing I need want to fix is the very syntactical approach to matching library calls. This is blocked on having a proper replacement for `Value::named("json.loads").getACall()` (so we can also recognize `from json import loads; loads(tainted_string)`).

### CODE REVIEW

**Comment by RasmusWL** [comment]
should update now that https://github.com/github/codeql/pull/4161 has been merged

**Comment by RasmusWL** [comment]
@yoff no changes after merging in #4161 -- looking over the PR, it looks like there was no test of flow _from_ a PostUpdateNote. So we need some more work on PostUpdateNote, but it's not blocking this PR as I see it.

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_collections.py
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_collections.py:81]:
```suggestion
    for k, v in tainted_dict.items():
```

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py:31]: Is `StringIO` implemented?

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py
Comment (yoff) [inline python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py:63]: This test passes, should it fail by the comment? (does the code represent 2 or 3 nested levels?)

**Review by yoff — CHANGES_REQUESTED** [review]
It is currently a little inconsistent if there are comments in the test code when the result is an expected failure.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:199]: yes :facepalm:

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_unpacking.py]: this is 3 levels `[[[`

File: python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py
Comment (RasmusWL) [reply python/ql/test/experimental/dataflow/tainttracking/defaultAdditionalTaintStep/test_json.py:31]: `StringIO` is not modeled currently.

**Comment by RasmusWL** [comment]
@yoff thanks for thorough review. I see that I forgot to update a lot of the comments for the tests based on the new implementataion :grimacing: but I fixed all the cases you pointed out now :+1:

**Review by yoff — COMMENTED** [review]
All changes look good, thanks for also adding the extra test 👍

**Review by yoff — APPROVED** [review]

**Review by tausbn — CHANGES_REQUESTED** [review]
Generally this looks good to me, but I have a few comments about how the code is structured.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:160]: Thea idea was that all these matches on `call.getFunction().(NameNode).getId()` are temporary until we have a proper replacement for `Value::named()`. It's not something I'm too concerned about fixing up. If it's something you feel very strongly about, I can give it a try :+1:

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:182]: Personally I don't find it too problematic, so how strong is your strong urge? :blush:

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll
Comment (tausbn) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:160]: Nah, I'm in the lower half of the sliding scale, given the context on this one. If we start adding a bunch more cases like these (pre-replacement), I definitely think we _should_ factor it out, if nothing else to make the replacement easier to implement.

File: python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll
Comment (tausbn) [reply python/ql/src/experimental/dataflow/internal/TaintTrackingPrivate.qll:182]: Not _that_ strong. :slightly_smiling_face:

**Review by tausbn — APPROVED** [review]

**Pull request merged**
